### PR TITLE
feat: missing actions added, ui improved

### PR DESCRIPTION
**Pull request by dokolyski**

### CODE REVIEW

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
+    <h1>TODO List</h1>
+    <button mat-raised-button color='primary' (click)='openCreateDialog()'>New task</button>
+  </todo-workspace-tasks-list>
+
+  <todo-workspace-tasks-list class='task-list-container__task-list  mat-card mat-elevation-z3'
+                             (deleteEvent)='tasksFacade.delete($event._id)'
+                             (editEvent)='openEditDialog($event)'
+                             (changePage)='tasksFacade.changePageDone($event.event, $event.tasksNumber)'
+                             (completionChangeEvent)='tasksFacade.changeCompletion($event)'
+                             [tasks]='tasks$ | complete: true | artificialPaginator: donePage$ | async'
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html]: for better optimisation, you can fetch data in your typescript, and then you don't need to use async twice.

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 2]
```diff
@@ -1,30 +1,66 @@
...
   templateUrl: './tasks-feature.component.html',
   styleUrls: ['./tasks-feature.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class TasksFeatureComponent implements OnInit {
   tasks$: Observable<Task[]>;
-  constructor(private tasksFacade: TasksFacade, private matDialog: MatDialog) {}
+  todoPage$: Observable<PaginatorState>;
+  donePage$: Observable<PaginatorState>;
+  loading$: Observable<boolean>;
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: `tasks$ = his.tasksFacade.tasks$;`

you don't need to set these variables in OnInit
then you can add `readonly` property

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 3]
```diff
@@ -1,30 +1,66 @@
...
+    this.subscription = this.tasksFacade.error$.subscribe((value) => {
+      this.snackbar.open(value);
+    });
   }
 
   openCreateDialog() {
+    this.openDialog('Add new task:', this.tasksFacade.create);
+  }
+
+  openEditDialog(task: Task) {
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts:59]: do you need two functions to `openDialog`?

File: libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+@Component({
+  selector: 'todo-workspace-progress-spinner-overlay',
+  templateUrl: './progress-spinner-overlay.component.html',
+  styleUrls: ['./progress-spinner-overlay.component.scss'],
+})
+export class ProgressSpinnerOverlayComponent implements OnInit {
+  constructor() {}
+
+  ngOnInit(): void {}
```
Comment (kstepien3) [inline libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts]: remove unused elements

File: libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'todo-workspace-progress-spinner-overlay',
+  templateUrl: './progress-spinner-overlay.component.html',
+  styleUrls: ['./progress-spinner-overlay.component.scss'],
+})
```
Comment (kstepien3) [inline libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts:8]: add Detection strategy onPush

File: libs/tasks/ui-task-form-dialog/src/lib/component/task-form-dialog.component.scss [hunk 6]
```diff
@@ -0,0 +1,11 @@
+.container {
+  max-width: 500px;
+}
+
+.container__form {
+  width: 100%;
+}
```
Comment (kstepien3) [inline libs/tasks/ui-task-form-dialog/src/lib/component/task-form-dialog.component.scss]:
```
.container__form {
  width: 100%;
  & > * {
    width:100%
  }
}
```

File: libs/tasks/ui-task-form-dialog/src/lib/component/task-form-dialog.component.scss [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+.container__form > * {
```
Comment (kstepien3) [inline libs/tasks/ui-task-form-dialog/src/lib/component/task-form-dialog.component.scss]: and remove it

File: libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss [hunk 8]
```diff
@@ -0,0 +1,31 @@
+.task-list__header {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  width: 100%;
+  height: 40px;
+}
+
+::ng-deep.task-list__header > * {
```
Comment (kstepien3) [inline libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss]: don't use `::ng-deep` this is _deprecated_

File: libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  margin: 0!important;
```
Comment (kstepien3) [inline libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss]: I think, `!important` is not necessary

**Review by kstepien3 — CHANGES_REQUESTED** [review]

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html (hunk 1 above)
Comment (dokolyski) [reply libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html]: ok

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts (hunk 3 above)
Comment (dokolyski) [reply libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts:59]: no, but openEditDialog is called in two places, and then I should pass the same parameters two times, so I though that this more specific functions could be useful

File: libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts (hunk 4 above)
Comment (dokolyski) [reply libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts]: ok

File: libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts (hunk 5 above)
Comment (dokolyski) [reply libs/tasks/ui-progress-spinner-overlay/src/lib/component/progress-spinner-overlay.component.ts:8]: ok

File: libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss (hunk 8 above)
Comment (dokolyski) [reply libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss]: ok

File: libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss (hunk 9 above)
Comment (dokolyski) [reply libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.scss]: yep

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html [hunk 10]
```diff
@@ -1,21 +1,27 @@
...
-<!--               [pageSizeOptions]='pageSizeOptions'-->
-<!--               (page)='pageEvent = $event'>-->
-<!--</mat-paginator>-->
+<div class='task-list-container'>
+  <todo-workspace-tasks-list class='task-list-container__task-list  mat-card mat-elevation-z3'
+                             (deleteEvent)='tasksFacade.delete($event._id)'
+                             (editEvent)='openEditDialog($event)'
+                             (changePage)='tasksFacade.changePageTodo($event.event, $event.tasksNumber)'
+                             (completionChangeEvent)='tasksFacade.changeCompletion($event)'
+                             [tasks]='tasks | complete: false | artificialPaginator: todoPage$ | async'
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html]: you can add `as todoPage` property. Like:
`todoPage$ | async as todoPage`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
+                             [pagination]='todoPage$ | async'>
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html]: then you don't need to use async. If you used  property `as todoPage` now you can provide here `todoPage` without `async`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html [hunk 12]
```diff
@@ -1,21 +1,27 @@
...
+    <h1 class='task-list-container__task-list__title'>TODO List</h1>
+    <button mat-raised-button color='primary' (click)='openCreateDialog()'>New task</button>
+  </todo-workspace-tasks-list>
+
+  <todo-workspace-tasks-list class='task-list-container__task-list  mat-card mat-elevation-z3'
+                             (deleteEvent)='tasksFacade.delete($event._id)'
+                             (editEvent)='openEditDialog($event)'
+                             (changePage)='tasksFacade.changePageDone($event.event, $event.tasksNumber)'
+                             (completionChangeEvent)='tasksFacade.changeCompletion($event)'
+                             [tasks]='tasks | complete: true | artificialPaginator: donePage$ | async'
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.html]: same. `donePage$ | async as donePage` and at line 19 set `donePage` value

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 13]
```diff
@@ -1,30 +1,77 @@
...
+} from '@angular/core';
+import {
+  PaginatorState,
+  Task,
+  TasksFacade,
+} from '@todo-workspace/tasks/data-access';
+import { Observable, Subscription } from 'rxjs';
 import { MatDialog } from '@angular/material/dialog';
 import { TaskFormDialogComponent } from '@todo-workspace/tasks/ui-task-form-dialog';
+import { SnackbarService } from '../../../../data-access/src/lib/services/snackbar.service';
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: wrong path. I think it should be started with `@todo`

File: libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.ts [hunk 14]
```diff
@@ -9,4 +17,25 @@ import { Task } from '@todo-workspace/tasks/data-access';
...
 export class TasksListComponent {
   @Input() tasks: Task[];
+  @Input() pagination: PaginatorState;
+  @Input() checked = false;
+  @Output() completionChangeEvent = new EventEmitter<Task>();
+  @Output() editEvent = new EventEmitter<Task>();
+  @Output() deleteEvent = new EventEmitter<Task>();
+  @Output() changePage = new EventEmitter<{
+    event: PageEvent;
+    tasksNumber: number;
```
Comment (kstepien3) [inline libs/tasks/ui-tasks-list/src/lib/component/tasks-list.component.ts]: maybe an interface should be a better way

File: proxy.config.json [hunk 15]
```diff
@@ -4,7 +4,7 @@
     "secure": true,
     "changeOrigin": true,
     "pathRewrite": {
-      "^/api": "/api/1d59cef4373c4cda8ee3f784bb0c4ef3"
+      "^/api": "/api/382dec296f2f4a3995fa4ef638a8c07c"
```
Comment (kstepien3) [inline proxy.config.json]: you can try to install `dotenv` and provide in there an API value. See if it makes any sense at all to make it work.

Also, add to README.md information about API. How to start an application with the generated API key.

**Review by kstepien3 — CHANGES_REQUESTED** [review]

File: proxy.config.json (hunk 15 above)
Comment (dokolyski) [reply proxy.config.json]: i saved the key value in environment var and used an interceptor to add it to url, kept proxy only to origin change

File: apps/todo-app/src/environments/environment.ts [hunk 16]
```diff
@@ -4,6 +4,8 @@
 
 export const environment = {
   production: false,
+  // api: 'dwdadaw',
```
Comment (kstepien3) [inline apps/todo-app/src/environments/environment.ts]: do you need this?

File: apps/todo-app/src/app/interceptors/api-key.interceptor.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+import { Observable } from 'rxjs';
+import { environment } from '../../environments/environment';
+
+@Injectable()
+export class ApiKeyInterceptor implements HttpInterceptor {
+  intercept(
+    request: HttpRequest<unknown>,
+    next: HttpHandler
+  ): Observable<HttpEvent<unknown>> {
+    const endpoint = request.url.split('/api/');
```
Comment (kstepien3) [inline apps/todo-app/src/app/interceptors/api-key.interceptor.ts]: why you split the endpoint?

File: jest.config.js [hunk 18]
```diff
@@ -5,5 +5,7 @@ module.exports = {
     '<rootDir>/libs/tasks/data-access',
     '<rootDir>/libs/tasks/ui-component',
     '<rootDir>/libs/tasks/ui-component',
```
Comment (kstepien3) [inline jest.config.js]: duplicated code

File: libs/tasks/data-access/src/lib/+state/tasks.actions.ts [hunk 19]
```diff
@@ -1,5 +1,6 @@
 import { createAction, props } from '@ngrx/store';
-import { Task } from '../resources/models/task';
+import { PaginatorState, Task } from '@todo-workspace/tasks/domain';
```
Comment (Kordrad) [inline libs/tasks/data-access/src/lib/+state/tasks.actions.ts:2]: Pagination state should be in a state scope

File: libs/tasks/data-access/src/lib/+state/tasks.actions.ts [hunk 20]
```diff
@@ -32,3 +35,8 @@ export const taskDelete = createAction(
   '[Tasks] Delete Task',
   props<{ id: number }>()
 );
+
+export const taskChangeCompletion = createAction(
+  '[Tasks] Change Task Completion',
+  props<{ task: Task }>()
```
Comment (kstepien3) [inline libs/tasks/data-access/src/lib/+state/tasks.actions.ts]: you can remove brackets

File: libs/tasks/data-access/src/lib/+state/tasks.facade.ts [hunk 21]
```diff
@@ -25,7 +35,45 @@ export class TasksFacade {
...
+    this.store$.dispatch(TasksActions.taskEdit({ task }));
+  }
+
+  delete(taskId: number) {
+    this.store$.dispatch(TasksActions.taskDelete({ id: taskId }));
+  }
+
+  changeCompletion(task: Task) {
+    this.store$.dispatch(
+      TasksActions.taskEdit({ task: { ...task, completed: !task.completed } })
```
Comment (kstepien3) [inline libs/tasks/data-access/src/lib/+state/tasks.facade.ts]: then you can also remove it from here

File: libs/tasks/data-access/src/lib/services/tasks-data.service.spec.ts [hunk 22]
```diff
@@ -1,16 +1,42 @@
...
+    });
     service = TestBed.inject(TasksDataService);
   });
 
   it('should be created', () => {
     expect(service).toBeTruthy();
   });
+
+  it('#getAll should return all records sorted by dueDate', () => {
+    // const httpGetMock: jest.Mock = jest.fn();
```
Comment (kstepien3) [inline libs/tasks/data-access/src/lib/services/tasks-data.service.spec.ts]: If you want to make something in the future please add `// Todo:` and describe it.

File: libs/tasks/domain/src/lib/tasks-domain.module.ts [hunk 23]
```diff
@@ -0,0 +1,7 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+
+@NgModule({
+  imports: [CommonModule],
+})
+export class TasksDomainModule {}
```
Comment (kstepien3) [inline libs/tasks/domain/src/lib/tasks-domain.module.ts]: remove this file if you don't import module in any place.

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 24]
```diff
@@ -1,30 +1,74 @@
...
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class TasksFeatureComponent implements OnInit {
-  tasks$: Observable<Task[]>;
-  constructor(private tasksFacade: TasksFacade, private matDialog: MatDialog) {}
+export class TasksFeatureComponent implements OnInit, OnDestroy {
+  tasks: Task[] = [];
+  readonly todoPage$: Observable<PaginatorState> = this.tasksFacade.todoPage$;
+  readonly donePage$: Observable<PaginatorState> = this.tasksFacade.donePage$;
+  readonly loading$: Observable<boolean> = this.tasksFacade.loading$;
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: you don't need set type when you set value

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 25]
```diff
@@ -1,30 +1,74 @@
...
+  errorSubscription: Subscription;
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: Visit this page: https://stackoverflow.com/questions/58428791/angular-rxjs-observable-takeuntil-vs-unsubscribe-with-a-subscription
then you can set variable like this:
`readonly errorSubscription$ = this.tasksFacade.error$`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts (hunk 24 above)
Comment (kstepien3) [reply libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: `readonly` variables should be at bottom of your list of variables

**Review by kstepien3 — CHANGES_REQUESTED** [review]

File: apps/todo-app/src/app/interceptors/api-key.interceptor.ts (hunk 17 above)
Comment (dokolyski) [reply apps/todo-app/src/app/interceptors/api-key.interceptor.ts]: this was for adding api key into url, I refactored it to be more readable

File: libs/tasks/data-access/src/lib/services/tasks-data.service.spec.ts [hunk 26]
```diff
@@ -1,16 +1,56 @@
...
+      providers: [{ provide: http }],
+    });
     service = TestBed.inject(TasksDataService);
   });
 
   it('should be created', () => {
     expect(service).toBeTruthy();
   });
+
+  it('#getAll should return all records sorted by dueDate', () => {
```
Comment (kstepien3) [inline libs/tasks/data-access/src/lib/services/tasks-data.service.spec.ts:31]: do not pass this test. You can use `it.skip`. Then you can remove comments from the bottom, but still use `// Todo` comment.

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 27]
```diff
@@ -1,30 +1,78 @@
...
 })
-export class TasksFeatureComponent implements OnInit {
-  tasks$: Observable<Task[]>;
-  constructor(private tasksFacade: TasksFacade, private matDialog: MatDialog) {}
+export class TasksFeatureComponent implements OnInit, OnDestroy {
+  tasks: Task[] = [];
+  readonly destroy$: Subject<any> = new Subject<any>();
+  readonly todoPage$: Observable<PaginatorState> = this.tasksFacade.todoPage$;
+  readonly donePage$: Observable<PaginatorState> = this.tasksFacade.donePage$;
+  readonly loading$: Observable<boolean> = this.tasksFacade.loading$;
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: you don't need set types if you set value.

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 28]
```diff
@@ -1,30 +1,78 @@
...
   templateUrl: './tasks-feature.component.html',
   styleUrls: ['./tasks-feature.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class TasksFeatureComponent implements OnInit {
-  tasks$: Observable<Task[]>;
-  constructor(private tasksFacade: TasksFacade, private matDialog: MatDialog) {}
+export class TasksFeatureComponent implements OnInit, OnDestroy {
+  tasks: Task[] = [];
+  readonly destroy$: Subject<any> = new Subject<any>();
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: `new Subject<void>()`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 29]
```diff
@@ -1,30 +1,78 @@
...
+  readonly loading$: Observable<boolean> = this.tasksFacade.loading$;
+
+  constructor(
+    public tasksFacade: TasksFacade,
+    private matDialog: MatDialog,
+    private snackbar: SnackbarService
+  ) {}
+
+  ngOnDestroy(): void {
+    this.destroy$.next({});
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: remove brackets

**Review by kstepien3 — CHANGES_REQUESTED** [review]

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 30]
```diff
@@ -1,30 +1,78 @@
...
+
+  private openDialog(
+    title: string,
+    onSuccessMethod: (this: TasksFacade, Task) => void,
+    task?: Task
+  ) {
     this.matDialog
-      .open(TaskFormDialogComponent, { data: { title: 'Add new task:' } })
+      .open(TaskFormDialogComponent, {
+        data: { title: title, task },
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: `data: { title, task },`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 31]
```diff
@@ -1,30 +1,78 @@
...
+      })
       .afterClosed()
       .subscribe((value: Task) => {
         if (value != null) {
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: `!!value`

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 32]
```diff
@@ -1,30 +1,78 @@
...
+
+    this.tasksFacade.error$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((value) => {
+        this.snackbar.open(value);
+      });
+
+    this.tasksFacade.tasks$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((value) => (this.tasks = value));
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: generally, you can change the `value` name to the `tasks` name. It is more readable

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 33]
```diff
@@ -1,30 +1,78 @@
...
+    this.destroy$.unsubscribe();
+  }
 
   ngOnInit(): void {
     this.tasksFacade.init();
-    this.tasks$ = this.tasksFacade.tasks$;
+
+    this.tasksFacade.error$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((value) => {
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: for example. Here I don't know what value you provide. Message, title, number?

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts [hunk 34]
```diff
@@ -1,30 +1,78 @@
...
+    onSuccessMethod: (this: TasksFacade, Task) => void,
+    task?: Task
+  ) {
     this.matDialog
-      .open(TaskFormDialogComponent, { data: { title: 'Add new task:' } })
+      .open(TaskFormDialogComponent, {
+        data: { title: title, task },
+      })
       .afterClosed()
       .subscribe((value: Task) => {
```
Comment (kstepien3) [inline libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: same, `task: Task` and in your code, I don't need to go back here to read what value you receive.

**Review by kstepien3 — CHANGES_REQUESTED** [review]

File: libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts (hunk 31 above)
Comment (dokolyski) [reply libs/tasks/feature/src/lib/tasks-feature/tasks-feature.component.ts]: eslint told that !! is redundant so i changed to if(value)

**Review by kstepien3 — APPROVED** [review]

**Pull request merged**